### ISSUE TITLE
Improving PS5 Integration: Attribute List or Auto-Entity Creation

### ISSUE THREAD

**Issue by Levtos** [JackJPowell/hass-psn#13]
**Is your feature request related to a problem? Please describe.**

As a Home Assistant beginner, I’m always frustrated when I can’t fully utilize all the attributes of the PS5 integration. I can see the game covers and trophy information, but it’s tough for me to access and integrate all the extra details through the helper.

**Describe the solution you'd like**

I’d love to have a comprehensive list of all available attributes, maybe even grouped by which entity provides which info. Alternatively, it would be awesome if the integration could automatically create all entities in Home Assistant, so I don’t have to hunt for them.

**Describe alternatives you've considered**

I’ve considered looking through documentation and existing issues, but it’s quite overwhelming, especially as I’m just getting started. Manual integration of each attribute also seems tedious and time-consuming.

**Additional context**

I’m really excited about using this integration, as it’s something I’ve been looking for for years! If you need help translating the integration into German, feel free to reach out!

**Comment by xXREDXIIIIXx** [comment]
Below you can find the current Entity IDs and the attributes associated with them, I sympathise with you about being a new user and everything can be overwhelming, you should find the HA community to be very welcoming with answering questions or pointing you to relevant documentation

> Manual integration of each attribute also seems tedious and time-consuming
 
  Jack has done a lot of work to bring us this, you should be prepared to learn how to use it.  there is a issue post from another user asking how to get the attributes #9, I have a post in there detailing how to do this.

**media_player.ps5_console** - Shows the title of the game you are playing or other information such as "Browsing the menu"

**sensor.psn_status** - Shows your online state - Online, Offline, Playing
- _The below information is only available when a game is open on the console left is the attribute name and right is the example output_
- Name - The First Descendant
- Description - A free-to-play, cross-platform, third-person looter shooter.
- Platform - PS5
- Content rating - ESRB Mature
- Play count - 42
- Play duration - 3 Days 2h
- Star rating - 4.02
- Trophies - bronze: 6, silver: 12, gold: 6, platinum: 1
- Earned trophies - bronze: 6, silver: 12, gold: 6, platinum: 1
- Trophy progress - 100

**sensor.playstation_trophy_level** Shows Current trophy Level - 651
- Platinum - 381
- Gold - 1,967
- Silver - 4,328
- Bronze - 10,584
- Next level progress - 10

I hope this helps you out, but feel free to ask, I don't normally check here often but I should get notified of any replies.

**Comment by Levtos** [comment]
Hey xXREDXIIIIXx,

Thanks for your reply and the helpful information! I tried to implement what you suggested this morning because I saw the code in issue #9, but I'm running into some errors. I added the following code to my `configuration.yaml`:


```
# PSN
template:
  - sensor:
      - name: "PSN Trophäen Level"
        unique_id: "psn_trophy_level"
        icon: "mdi:trophy"
        state: "{{ state_attr('sensor.psn_data', 'trophy_summary').trophy_level }}"
        attributes:
          platinum: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['platinum'] }}"
          gold: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['gold'] }}"
          silver: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['silver'] }}"
          bronze: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['bronze'] }}"
          next_level_progress: "{{ state_attr('sensor.psn_data', 'trophy_summary').progress }}"
          
      - name: "PSN Status"
        unique_id: "psn_status"
        icon: "mdi:account-circle-outline"
        state: "{{ state_attr('sensor.psn_data', 'status') }}"
        attributes:
          name: "{{ state_attr('sensor.psn_data', 'status')['name'] }}"
          description: "{{ state_attr('sensor.psn_data', 'status')['description'] }}"
          platform: "{{ state_attr('sensor.psn_data', 'status')['platform'] }}"
          content_rating: "{{ state_attr('sensor.psn_data', 'status')['content_rating'] }}"
          play_count: "{{ state_attr('sensor.psn_data', 'status')['play_count'] }}"
          play_duration: "{{ state_attr('sensor.psn_data', 'status')['play_duration'] }}"
          platinum_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['platinum'] }}"
          gold_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['gold'] }}"
          silver_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['silver'] }}"
          bronze_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['bronze'] }}"
          earned_platinum_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['platinum'] }}"
          earned_gold_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['gold'] }}"
          earned_silver_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['silver'] }}"
          earned_bronze_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['bronze'] }}"
          trophy_progress: "{{ state_attr('sensor.psn_data', 'status')['trophy_progress'] }}"
```

Unfortunately, I'm encountering some errors. If you could provide any insights or suggestions, that would be awesome!
Example Error:

`Logger: homeassistant.helpers.event
Quelle: helpers/template.py:621
Erstmals aufgetreten: 21:21:21 (12 Vorkommnisse)
Zuletzt protokolliert: 21:21:21

Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['trophies']['bronze'] }}) renders=2>
Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['platinum'] }}) renders=2>
Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['gold'] }}) renders=2>
Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['silver'] }}) renders=2>
Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['bronze'] }}) renders=2>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 619, in async_render
    render_result = _render_with_context(self.template, compiled, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2654, in _render_with_context
    return template.render(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 939, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
  File "/usr/local/lib/python3.12/site-packages/jinja2/sandbox.py", line 304, in getitem
    return obj[argument]
           ~~~^^^^^^^^^^
jinja2.exceptions.UndefinedError: 'None' has no attribute 'earned_trophies'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 731, in async_render_to_info
    render_info._result = self.async_render(  # noqa: SLF001
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 621, in async_render
    raise TemplateError(err) from err
homeassistant.exceptions.TemplateError: UndefinedError: 'None' has no attribute 'earned_trophies'`

Even when I manually add the helper under Settings -> Devices & Integrations -> Helpers, I still get the similiar error messages. Could it be possible that there's an issue with the translations since I'm logged in via de-DE?

Thanks again for your help!

Best,
Levtos

**Comment by xXREDXIIIIXx** [comment]
> Hey xXREDXIIIIXx,
> 
> Thanks for your reply and the helpful information! I tried to implement what you suggested this morning because I saw the code in issue #9, but I'm running into some errors. I added the following code to my `configuration.yaml`:
> 
> ```
> # PSN
> template:
>   - sensor:
>       - name: "PSN Trophäen Level"
>         unique_id: "psn_trophy_level"
>         icon: "mdi:trophy"
>         state: "{{ state_attr('sensor.psn_data', 'trophy_summary').trophy_level }}"
>         attributes:
>           platinum: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['platinum'] }}"
>           gold: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['gold'] }}"
>           silver: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['silver'] }}"
>           bronze: "{{ state_attr('sensor.psn_data', 'trophy_summary').earned_trophies['bronze'] }}"
>           next_level_progress: "{{ state_attr('sensor.psn_data', 'trophy_summary').progress }}"
>           
>       - name: "PSN Status"
>         unique_id: "psn_status"
>         icon: "mdi:account-circle-outline"
>         state: "{{ state_attr('sensor.psn_data', 'status') }}"
>         attributes:
>           name: "{{ state_attr('sensor.psn_data', 'status')['name'] }}"
>           description: "{{ state_attr('sensor.psn_data', 'status')['description'] }}"
>           platform: "{{ state_attr('sensor.psn_data', 'status')['platform'] }}"
>           content_rating: "{{ state_attr('sensor.psn_data', 'status')['content_rating'] }}"
>           play_count: "{{ state_attr('sensor.psn_data', 'status')['play_count'] }}"
>           play_duration: "{{ state_attr('sensor.psn_data', 'status')['play_duration'] }}"
>           platinum_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['platinum'] }}"
>           gold_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['gold'] }}"
>           silver_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['silver'] }}"
>           bronze_trophies: "{{ state_attr('sensor.psn_data', 'status')['trophies']['bronze'] }}"
>           earned_platinum_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['platinum'] }}"
>           earned_gold_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['gold'] }}"
>           earned_silver_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['silver'] }}"
>           earned_bronze_trophies: "{{ state_attr('sensor.psn_data', 'status')['earned_trophies']['bronze'] }}"
>           trophy_progress: "{{ state_attr('sensor.psn_data', 'status')['trophy_progress'] }}"
> ```
> 
> Unfortunately, I'm encountering some errors. If you could provide any insights or suggestions, that would be awesome! Example Error:
> 
> ```
> `Logger: homeassistant.helpers.event
> Quelle: helpers/template.py:621
> Erstmals aufgetreten: 21:21:21 (12 Vorkommnisse)
> Zuletzt protokolliert: 21:21:21
> 
> Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['trophies']['bronze'] }}) renders=2>
> Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['platinum'] }}) renders=2>
> Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['gold'] }}) renders=2>
> Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['silver'] }}) renders=2>
> Error while processing template: Template<template=({{ state_attr('sensor.psn_data', 'status')['earned_trophies']['bronze'] }}) renders=2>
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 619, in async_render
>     render_result = _render_with_context(self.template, compiled, **kwargs)
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2654, in _render_with_context
>     return template.render(**kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 1304, in render
>     self.environment.handle_exception()
>   File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 939, in handle_exception
>     raise rewrite_traceback_stack(source=source)
>   File "<template>", line 1, in top-level template code
>   File "/usr/local/lib/python3.12/site-packages/jinja2/sandbox.py", line 304, in getitem
>     return obj[argument]
>            ~~~^^^^^^^^^^
> jinja2.exceptions.UndefinedError: 'None' has no attribute 'earned_trophies'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 731, in async_render_to_info
>     render_info._result = self.async_render(  # noqa: SLF001
>                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 621, in async_render
>     raise TemplateError(err) from err
> homeassistant.exceptions.TemplateError: UndefinedError: 'None' has no attribute 'earned_trophies'`
> ```
> 
> Even when I manually add the helper under Settings -> Devices & Integrations -> Helpers, I still get the similiar error messages. Could it be possible that there's an issue with the translations since I'm logged in via de-DE?
> 
> Thanks again for your help!
> 
> Best, Levtos

Not 100% Sure but I think the state you are trying to get is invalid:

> state: "{{ state_attr('sensor.psn_data', 'trophy_summary').trophy_level }}"

Try the below as a test, these should just bring back the trophy level and your platinum count
{{ states('sensor.playstation_trophy_level') }}
{{ state_attr('sensor.playstation_trophy_level', 'platinum') }}

**Comment by JackJPowell** [comment]
@Levtos, I just pushed out a beta release of v0.6.0 that includes the option to expose all attributes as sensor entities. You can go into the HACS 2.0 integration to enable beta updates for the integration. Give it a shot and let me know how it goes.

**Comment by JackJPowell** [comment]
On my first attempt, I added the new entities to the existing setup via "Configure." I noticed that the old sensors follow the naming logic sensor.psn_entity, whereas the new ones are named sensor.psnname_entity. It’s not a major issue, just a small cosmetic detail ;)

**This is intentional and was done to support multiple psn accounts. Without including the psn name, people with multiple PSN accounts would generate the same entities for each causing a conflict and resulting in entities that end with _2.**

For the second attempt, I removed my existing console from the integration and reconnected via NPSSO. Here, the same naming inconsistency appears. As I said, it’s not a big deal since entities can be renamed as desired, but personally, I prefer a naming convention like sensor_integration_entity or sensor.device_entity. Again, this is just a cosmetic detail, and I’m not sure if it’s intended.

**Same as above. You held on to the old entity ID until you deleted and re-added everything. It's easy to set the entity ID as the developer, but the correct development method is to let HA derive the entity ID from the data provided. You can read about it and the has_entity_name attribute [here](https://developers.home-assistant.io/docs/core/entity/#has_entity_name-true-mandatory-for-new-integrations).**

On my third test, I reviewed the individual entities. To illustrate better, I created a custom dashboard and added all entities. The entity "earned Trophy" doesn’t seem to work properly. All four fields (Bronze, Silver, Gold, and Platinum) display the number 13. I’m not sure what this number means or where it comes from.
In my first test, I selected Diablo IV, where the number 13 appeared, and I observed the same in Overwatch 2 and Sid Meier’s Civ VI. The number 13 keeps appearing. I’ll send you two screenshots to demonstrate this, one of the dashboard and one of the trophy overview from my account (apologies in advance that both images are in German).

**Good find! I can't believe I missed this. Next update will fix this.**

Lastly, I noticed a fourth issue, though it’s not directly related to this one. I'll report a new issue for it within the next few hours.

**I think I know what's going on here. I'll check and respond to the issue**

**Comment by JackJPowell** [comment]
Thanks! I've got it fixed with some bonus stuff too. I'll have a new beta up shortly.

**Comment by JackJPowell** [comment]
Hey @Levtos, just pushed out beta 4. I hope this is the one to release

**Comment by JackJPowell** [comment]
Thanks @Levtos. I'm a bit surprised about the lack of "About me" text. It was showing up for me. 
The image I'm pulling is large so you may have to explicitly scale it to make it fit in the card. 
When I asked on the home assistant discord, the reply I got did say to set the value to None. I'm still not sure it's 100% right but I don't think I should set them to an empty string. I'm curious what value they would need to be for you to not need to handle it manually. 

I'm pushing out another beta right now that should at least address the text field issues.

**Comment by JackJPowell** [comment]
I've finally released v0.7. Thanks for all the testing!
